### PR TITLE
feat(pytorch): add mimo-v2-flash support

### PR DESCRIPTION
**Pull request by yidingcheng0206**
## Motivation

This PR adds support for Xiaomi MiMo-V2-Flash in LMDeploy's PyTorch backend. MiMo-V2-Flash uses hybrid full-attention and sliding-window-attention (SWA) layers and provides a three-layer MTP module for speculative decoding. This PR adds the required model implementation, attention and KV-cache support, MTP integration, and tool-call parsing.

## Modification

- Add the MiMo-V2-Flash model configuration and target model implementation.
- Add the three-layer MiMo MTP draft model.
- Add support for the model's hybrid full-attention/SWA layer pattern.
- Add SWA state-ring cache and related CUDA attention kernels.
- Add tensor-parallel KV-head replication for full-attention and SWA layers.
- Add paged SWA KV-cache handling for speculative decoding.
- Integrate MiMo MTP with LMDeploy speculative decoding and CUDA-graph paths.
- Add the MiMo tool-call parser and register it with the parser manager.
- Add the FP8 linear and attention paths required by the model.
- Update model registration, supported-model lists, and English/Chinese documentation.

LMDeploy and Hugging Face were evaluated with aligned settings. The LMDeploy values are averages over repeated runs.

| Benchmark        | LMDeploy              | Hugging Face | Difference |
| ---------------- | --------------------- | ------------ | ---------- |
| GPQA-Diamond     | 84.47% (8-run average) | 83.70%      | +0.77 pp   |
| AIME 2025        | 95.10% (32-run average) | 94.10%     | +1.00 pp   |

## Acceptance of MTP

The following results use TP8 + DP2 + EP8, attention TP4 per DP group, fixed 128-request ShareGPT workload, 2048 generated tokens, temperature 0, `ignore_eos=true`, three MTP draft positions, and three repeated runs.

### Aggregate results

| Engine   | Mean accept length | Acceptance rate                          |
| -------- | ------------------ | ---------------------------------------- |
| LMDeploy | 3.115              | 70.49% (native draft-token definition)   |
| SGLang   | 3.076              | 76.91% (including bonus-token normalization) |

### Per-position results

| Engine   | Position 1 | Position 2 | Position 3 | Strict 3/3 acceptance |
| -------- | ---------- | ---------- | ---------- | --------------------- |
| LMDeploy | 85.89%     | 69.38%     | 56.20%     | 56.20%                |
| SGLang   | 84.86%     | 67.99%     | 54.78%     | 54.78%                |

LMDeploy's native acceptance rate is `accepted_draft_tokens / drafted_tokens` and excludes the bonus token. Mean accept length includes the bonus token. Position values are prefix acceptance probabilities; they must not be multiplied together. The strict 3/3 rate is the probability that all three draft positions are accepted.

## Performance

The fixed-output benchmark uses the same ShareGPT inputs, TP8 + DP2 + EP8 topology, attention TP4 per DP group, concurrency 128, and three repetitions per point.

### MTP throughput and TTFT

| Engine         | Output | Duration (s) | Request throughput (req/s) | Output throughput (tok/s) | TTFT Mean (ms) | TTFT P50 (ms) | TTFT P99 (ms) |
| -------------- | ------ | ------------ | -------------------------- | ------------------------- | -------------- | ------------- | ------------- |
| LMDeploy MTP   | 2048   | 36.46        | 3.51                       | 7195.01                   | 934.35         | 909.23        | 1029.04       |
| SGLang MTP     | 2048   | 35.18        | 3.64                       | 7457.21                   | 469.01         | 463.74        | 535.28        |
| LMDeploy MTP   | 4096   | 66.54        | 1.93                       | 7896.77                   | 689.53         | 676.68        | 797.08        |
| SGLang MTP     | 4096   | 63.66        | 2.01                       | 8239.97                   | 366.37         | 367.62        | 389.05        |
| LMDeploy MTP   | 8192   | 121.75       | 1.06                       | 8677.01                   | 733.45         | 731.26        | 798.63        |
| SGLang MTP     | 8192   | 119.26       | 1.08                       | 8806.60                   | 692.20         | 701.28        | 739.12        |

### MTP TPOT and ITL

| Engine         | Output | TPOT Mean (ms) | TPOT P50 (ms) | TPOT P99 (ms) | ITL Mean (ms) | ITL P50 (ms) | ITL P99 (ms) |
| -------------- | ------ | -------------- | ------------- | ------------- | ------------- | ------------ | ------------ |
| LMDeploy MTP   | 2048   | 13.21          | 12.93         | 16.96         | 41.75         | 38.39        | 180.02       |
| SGLang MTP     | 2048   | 13.44          | 13.42         | 16.83         | 41.20         | 38.49        | 126.80       |
| LMDeploy MTP   | 4096   | 12.08          | 11.84         | 15.74         | 41.19         | 38.62        | 118.37       |
| SGLang MTP     | 4096   | 12.27          | 12.15         | 15.26         | 40.28         | 38.60        | 58.35        |
| LMDeploy MTP   | 8192   | 11.32          | 11.20         | 13.57         | 41.12         | 38.96        | 75.59        |
| SGLang MTP     | 8192   | 11.61          | 11.44         | 13.95         | 40.72         | 39.08        | 49.32        |

LMDeploy MTP output throughput is 96.5%, 95.8%, and 98.5% of SGLang MTP for 2048, 4096, and 8192 output tokens, respectively. MTP TPOT remains within 3% of SGLang in these tests.

## Runtime validation

- TP4 and TP8 target inference and CUDA Graph serving.
- TP8 + DP2 + EP8 + DeepEP + DeepGEMM + MTP3 production topology.
- Prefix caching with MTP and high-concurrency replay.
- SWA boundary and speculative accept/reject/rollback cases.
- MiMo tool-call API round trips.

## Compatibility

No intentional backward-incompatible changes are introduced. MiMo-V2-Flash support targets the PyTorch backend. MTP requires the three-layer draft weights included in the checkpoint. FA3 is optional; unsupported asymmetric head-dimension builds automatically use Triton.

## Validation

- Ruff and `git diff --check` pass.
- Focused model, attention, cache, MTP, FP8, distributed, CUDA Graph, and parser tests pass.
- Real-checkpoint TP4/TP8 and production-topology runtime validation pass.

## Assistance 

Assisted with Codex, reviewed manually.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: lmdeploy/pytorch/backends/cuda/graph_runner/runner.py [hunk 1]
```diff
@@ -252,12 +265,40 @@ def _capture_full_graph(self, graph_key: tuple, kwargs: dict[str, Any]):
...
-        output = runner.capture(**kwargs)
+        capture_cache = self.model.get_cudagraph_capture_cache(
+            kwargs['past_key_values'], spec_step_idx=int(kwargs.get('spec_step_idx', 0)))
+        block_ids: torch.Tensor | None = None
+        snapshot: list[torch.Tensor] | None = None
+        if capture_cache is not None:
+            attn_metadata = kwargs['attn_metadata']
+            batch_size = attn_metadata.q_seqlens.numel()
+            block_ids = torch.unique(attn_metadata.block_offsets[:batch_size]).long()
+            snapshot = [tensor.index_select(0, block_ids).clone() for tensor in capture_cache]
```
[the remark this answers is not in the document]
Comment (yidingcheng0206) [reply lmdeploy/pytorch/backends/cuda/graph_runner/runner.py]: The capture-cache snapshot path was introduced by this PR and previously passed raw block offsets to index_select and index_copy_. I now flatten and filter block IDs to the valid range [0, num_blocks).before snapshot/restore, and skip the cache snapshot when no valid IDs remain. The regression test covers negative, duplicate, and out-of-range block IDs.

File: lmdeploy/pytorch/kernels/cuda/pagedattention.py [hunk 2]
```diff
@@ -60,6 +60,7 @@ def _fwd_grouped_split_kernel(
     stride_boffb,
     kv_group_num: tl.constexpr,
     seq_len: tl.constexpr,
+    causal_multi_token: tl.constexpr,
```
Comment (grimoire) [inline lmdeploy/pytorch/kernels/cuda/pagedattention.py:63]: There are large change in this kernel, It could be risky to do such large change in this kernel, we better create a new kernel for swa.

File: lmdeploy/pytorch/backends/cuda/graph_runner/runner.py [hunk 3]
```diff
@@ -252,12 +253,35 @@ def _capture_full_graph(self, graph_key: tuple, kwargs: dict[str, Any]):
...
+        )
+        capture_state = self.model.get_cudagraph_capture_state(capture_context)
+        if capture_state is not None:
+            capture_state.snapshot()
+
+        try:
+            output = runner.capture(**kwargs)
+        finally:
+            if capture_state is not None:
+                capture_state.restore()
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/cuda/graph_runner/runner.py:273]: I am not sure if it is worth to add these restore logical. the cache might already been polluted during capture.

File: lmdeploy/pytorch/model_inputs.py [hunk 4]
```diff
@@ -233,6 +234,12 @@ class ModelInputs:
     logits_indices: torch.LongTensor | None = None
     # Number of compact logprob rows emitted for each sequence.
     seq_logit_length: torch.LongTensor | None = None
+    # Prediction-depth index for multi-layer MTP draft models.
```
Comment (grimoire) [inline lmdeploy/pytorch/model_inputs.py:237]: Do we need to set/update these values in strategy/model_inputs.py?
